### PR TITLE
Extract lib viewport

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: #5508

Depends on:
- #5644

A toolkit-neutral abstraction of a viewport, with horizontal and vertical scrollbars.

class Viewport relies on the abstract, façade class ViewportCallbacks which has it wxWidgets implementation in ProjectWindow.cpp.

There are now fewer dependencies on ProjectWindow.h.

QA: Unchanged zooming (including resizing of slider thumbs) and scrolling of the Track Panel after:
- [x] LOF (list-of-files) import with an "offset" command
- [x] Shift + mouse wheel
- [x] Ctrl (or Command) + mouse wheel
- [x] unmodified mouse wheel 
- [x] Dragging slider thumbs
- [x] Clicking arrowheads next to sliders
- [x] Clicking blank areas beside slider thumbs
- [x] Resizing main window
- [x] Page Up / Page Down keys
- [x] Deleting the last remaining track
- [x] Adding or removing tracks of any type (so that the vertical thumb appears, disappears, or resizes)
- [x] When a new tracks are pasted, or result from import, or generation, or recording, or analyzers (new label tracks), the first new one scrolls _completely_ into view vertically
- [x] Import of audio or MIDI also zooms to fit all, horizontally
- [x] Undo/Redo scrolls the first selected (or if none, the first) track completely into view vertically
- [x] Similar for up, down, Ctrl+Home, Ctrl+End
- [x] The previous again, with '"Move track focus" cycles repeatedly through tracks' checked in Tracks Behaviors preferences
- [x] Everything in Zoom, Track Size, Skip To sub-menus of View (except the checkbox item; that only involves track vertical rulers)
- [x] Zoom-in command during playback keeps the play position in view
- [x] Zoom to fit vertically scrolls to top of track list, in case there are too many tracks to fix
- [x] Tracks preferences:  vary Preset1 and Preset2, to test the "Zoom Toggle" command
- [x] Playback or recording, with or without pinning the play head
- [x] Clicking and dragging selection start/end off the left/right edge, causing horizontal scrolling
- [x] For completeness, try things with "Enable scrolling left of zero" in Tracks Behaviors, though that may be on the "kill list" for near future releases
- [x] Time ruler synchronizes with horizontal scrolling
- [x] Just open these windows and close, observing proper positioning:  Macro manager, macro palette, Frequency analysis, Timer record, Contrast, cross-document pasting dialog (with copied, hidden clip data), Export audio, Import Raw, Import Labels, Resample tracks dialog (which chooses rate), Edit Labels dialog
- [x] Test Screenshot macro command (just once, any settings); name of resulting file
- [x] Close a project window
- [x] Project repaints after changing the checkbox or drop-down in Snapping toolbar
- [x] Project repaints after clicking a line in the History dialog
- [x] Resize main window; then, Tools > Reset Configuration restores default size

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Comment by Paul-Licameli** [comment]
Rebased and self-reviewed and ready for review.  Also a QA checklist.

**Comment by Paul-Licameli** [comment]
Rebased onto recent master with conflict resolution (mostly just `#include`s)

**Comment by Paul-Licameli** [comment]
Another rebase to resolve conflicts, only in the Cmake files

**Review by crsib — APPROVED** [review]

File: libraries/lib-viewport/Viewport.h [hunk 1]
```diff
@@ -183,6 +183,23 @@ class VIEWPORT_API Viewport final
...
+   wxInt64 sbarTotal{ 1 };
+
+   // Internal wxScrollbar positions are only int in range, so multiply
+   // the above values with the following member to get the actual
+   // scroll bar positions as reported by the horizontal wxScrollbar's members
+   // i.e. units are scroll increments per pixel
+   double sbarScale{ 1.0 };
+
+   // Vertical scroll step
+   int scrollStep{ 16 };
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-viewport/Viewport.h:217]: Easily done.  Of course this was just cut-and-paste.

**Comment by Paul-Licameli** [comment]
Rebased unchanged to more recent master, but there are bugs to fix.

**Comment by Paul-Licameli** [comment]
> There's only one issue - the unmodified scroll wheel does not let the User scroll through the tracks vertically.

That problem bisects to the commit "More member functions moved from ProjectWindow to Viewport"

**Comment by Paul-Licameli** [comment]
Rebased onto recent master (which gained just one commit) and inserted a FIXUP commit for the silly little mistakes

@crsib a quick second round of review should do

@chinakov please retest, paying closer attention to the vertical scrollbar and thumb size.

**Review by crsib — APPROVED** [review]

**Pull request merged**